### PR TITLE
Support section header actions

### PR DESCRIPTION
**Pull request by wychoong**
- [x] Changes have been thoroughly tested to not break existing functionality.
- [x] New functionality has been documented or existing documentation has been updated to reflect changes.
- [x] Visual changes are explained in the PR description using a screenshot/recording of before and after.

infolist section
image

form section
image

wrapped
image

### CODE REVIEW

File: packages/support/resources/views/components/section/index.blade.php [hunk 1]
```diff
@@ -17,10 +18,15 @@
...
+    $actions = array_filter(
+        $actions,
+        fn ($action): bool => $action->isVisible(),
+    );
+    $hasActions = filled($actions);
     $hasDescription = filled((string) $description);
     $hasHeading = filled($heading);
     $hasIcon = filled($icon);
-    $hasHeader = $hasIcon || $hasHeading || $hasDescription || $collapsible || filled((string) $headerEnd);
+    $hasHeader = $hasActions || $hasIcon || $hasHeading || $hasDescription || $collapsible || filled((string) $headerEnd);
```
Comment (zepfietje) [inline packages/support/resources/views/components/section/index.blade.php]:
```suggestion
    $hasHeader =  $hasIcon || $hasHeading || $hasDescription || $collapsible || $hasActions || filled((string) $headerEnd);
```

File: packages/support/resources/views/components/section/index.blade.php [hunk 2]
```diff
@@ -17,10 +18,15 @@
...
+    $actions = array_filter(
+        $actions,
+        fn ($action): bool => $action->isVisible(),
+    );
+    $hasActions = filled($actions);
     $hasDescription = filled((string) $description);
     $hasHeading = filled($heading);
     $hasIcon = filled($icon);
-    $hasHeader = $hasIcon || $hasHeading || $hasDescription || $collapsible || filled((string) $headerEnd);
+    $hasHeader =  $hasIcon || $hasHeading || $hasDescription || $collapsible || $hasActions || filled((string) $headerEnd);
```
Comment (zepfietje) [inline packages/support/resources/views/components/section/index.blade.php]:
```suggestion
    $hasHeader = $hasIcon || $hasHeading || $hasDescription || $collapsible || $hasActions || filled((string) $headerEnd);
```

File: packages/support/resources/views/components/section/index.blade.php [hunk 3]
```diff
@@ -124,6 +130,12 @@
                 </div>
             @endif
 
+            @if ($hasActions)
+                <div class="flex flex-1 justify-end">
```
Comment (zepfietje) [inline packages/support/resources/views/components/section/index.blade.php]: We can probably remove the wrapper and directly add a class like `ms-auto` to the actions component used here?

**Review by zepfietje — CHANGES_REQUESTED** [review]
Thanks for the PR, @wychoong!

Could you have a look at my comment?
Also, @danharrin, don't we need some caching for the actions?

Additionally, maybe the API should be renamed to `headerActions`?

File: packages/support/resources/views/components/section/index.blade.php (hunk 3 above)
Comment (wychoong) [reply packages/support/resources/views/components/section/index.blade.php]: thats better, done!

Comment (wychoong) [reply packages/support/resources/views/components/section/index.blade.php]: the only difference I see is that now heading and action container is not taking half as no more `flex-1`
image

but it looks better as it doesnt wrap when larger than `md`
or do you think need to responsive based on viewport size

Comment (zepfietje) [reply packages/support/resources/views/components/section/index.blade.php]: Hmm but the heading wrapper has `flex-1`? 🤔

**Comment by wychoong** [comment]
> Also, @danharrin, don't we need some caching for the actions?

`getActions` is cached

> Additionally, maybe the API should be renamed to `headerActions`?

probably an alias to `actions`? or you thinking to support footerActions?

File: packages/support/resources/views/components/section/index.blade.php (hunk 3 above)
Comment (wychoong) [reply packages/support/resources/views/components/section/index.blade.php]: yup it does

**Comment by zepfietje** [comment]
Yeah potentially there will be footer actions in the future, so I'd prefer to have specific APIs.

**Comment by wychoong** [comment]
> Yeah potentially there will be footer actions in the future, so I'd prefer to have specific APIs.

I see. I think now only remain the api. I’ll leave it up to maintainers to decide for changes

**Comment by zepfietje** [comment]
Requesting a review from @danharrin for the API.
Note that I still want to review/rework the implementation in the view though.

**Comment by akunbeben** [comment]
Can't wait to have this feature 🎉

**Comment by zepfietje** [comment]
@wychoong, could you rework this to be `headerActions`?

File: packages/forms/src/Components/Concerns/HasHeaderActions.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+use Closure;
+use Filament\Forms\Components\Actions\Action;
+use Filament\Support\Enums\ActionSize;
+use Illuminate\Support\Arr;
+
+trait HasHeaderActions
+{
+    /**
+     * @var array<Action> | null
```
Comment (zepfietje) [inline packages/forms/src/Components/Concerns/HasHeaderActions.php:12]: This array could also contain action groups, right?

File: packages/forms/src/Components/Concerns/HasHeaderActions.php [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+     */
+    public function cacheHeaderActions(): array
+    {
+        $this->cachedHeaderActions = [];
+
+        foreach ($this->headerActions as $headerAction) {
+            foreach (Arr::wrap($this->evaluate($headerAction)) as $action) {
+                $this->cachedHeaderActions[$action->getName()] = $this->prepareAction(
+                    $action
+                        ->defaultSize(ActionSize::Small)
```
Comment (zepfietje) [inline packages/forms/src/Components/Concerns/HasHeaderActions.php]: Not sure about the default size yet.
Will check later on during the review.

File: packages/infolists/src/Components/Concerns/HasHeaderActions.php [hunk 6]
```diff
@@ -0,0 +1,69 @@
+<?php
```
Comment (zepfietje) [inline packages/infolists/src/Components/Concerns/HasHeaderActions.php:1]: Could this file be extracted for forms and infolists, or are there any differences between them?

File: packages/support/resources/views/components/section/index.blade.php [hunk 7]
```diff
@@ -3,6 +3,7 @@
 @endphp
 
 @props([
+    'headerActions' => [],
```
Comment (zepfietje) [inline packages/support/resources/views/components/section/index.blade.php]: Could you move this down to be in alphabetical order?

File: packages/forms/src/Components/Concerns/HasHeaderActions.php (hunk 4 above)
Comment (wychoong) [reply packages/forms/src/Components/Concerns/HasHeaderActions.php:12]: if not mistaken currently Form and Infolist does not support action groups
https://filamentphp.com/docs/3.x/forms/actions
https://filamentphp.com/docs/3.x/infolists/actions

only Table has it
https://filamentphp.com/docs/3.x/tables/actions#grouping-actions

File: packages/infolists/src/Components/Concerns/HasHeaderActions.php (hunk 6 above)
Comment (wychoong) [reply packages/infolists/src/Components/Concerns/HasHeaderActions.php:1]: its the same, can be extracted to Support
what would be a better name if to support footerActions in the future, like HasAffixes for prefix & suffix actions, or to rename it later when there is footerActions implementation

File: packages/forms/src/Components/Concerns/HasHeaderActions.php (hunk 4 above)
Comment (zepfietje) [reply packages/forms/src/Components/Concerns/HasHeaderActions.php:12]: Good point. I do think we need them for section actions to be useful to be honest, @danharrin?

File: packages/infolists/src/Components/Concerns/HasHeaderActions.php (hunk 6 above)
Comment (zepfietje) [reply packages/infolists/src/Components/Concerns/HasHeaderActions.php:1]: We can always refactor later. Focusing on just the header actions makes our lives simpler right now. 🙂

File: packages/forms/src/Components/Concerns/HasHeaderActions.php (hunk 4 above)
Comment (danharrin) [reply packages/forms/src/Components/Concerns/HasHeaderActions.php:12]: I think it would be quite a bit of work, lets hold off for now

**Comment by danharrin** [comment]
Please add some documentation

File: packages/infolists/src/Components/Concerns/HasHeaderActions.php (hunk 6 above)
Comment (wychoong) [reply packages/infolists/src/Components/Concerns/HasHeaderActions.php:1]: @zepfietje can you handle this extraction? not too sure how you would like to organize it in `Filament\Support\Concerns`

only thing to take note is action type `Filament\Forms\Components\Actions\Action` and `Filament\Infolists\Components\Actions\Action`

I believe the same need to be done for the interface as well

**Review by danharrin — APPROVED** [review]

File: docs-assets/app/app/Livewire/Forms/LayoutDemo.php [hunk 8]
```diff
@@ -300,6 +300,33 @@ public function form(Form $form): Form
...
+                Group::make()
+                    ->id('sectionHeaderActions')
+                    ->extraAttributes([
+                        'class' => 'p-16 max-w-2xl',
+                    ])
+                    ->schema([
+                        Section::make('Rate limiting')
+                            ->description('Prevent abuse by limiting the number of requests per period')
+                            ->headerActions([
+                                Action::make('test'),
```
Comment (zepfietje) [inline docs-assets/app/app/Livewire/Forms/LayoutDemo.php:312]: Shouldn't this also be "edit", @danharrin?
```suggestion
                                Action::make('edit'),
```

Comment (zepfietje) [reply docs-assets/app/app/Livewire/Forms/LayoutDemo.php:312]: Sorry, no, it makes sense as is.

File: docs-assets/screenshots/images/light/forms/layout/section/actions.jpg
Comment (zepfietje) [inline docs-assets/screenshots/images/light/forms/layout/section/actions.jpg:1]: This should be light theme, @danharrin?

File: docs-assets/screenshots/images/light/infolists/layout/section/actions.jpg
Comment (zepfietje) [inline docs-assets/screenshots/images/light/infolists/layout/section/actions.jpg:1]: This should be light too, @danharrin?

File: packages/support/resources/views/components/section/index.blade.php [hunk 9]
```diff
@@ -124,6 +130,13 @@
                 </div>
             @endif
 
+            @if ($hasHeaderActions)
+                <x-filament-actions::actions
+                    :actions="$headerActions"
```
Comment (zepfietje) [inline packages/support/resources/views/components/section/index.blade.php:135]: Maybe this should have alignment end?

File: packages/infolists/src/Components/Concerns/HasHeaderActions.php [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+     */
+    public function cacheHeaderActions(): array
+    {
+        $this->cachedHeaderActions = [];
+
+        foreach ($this->headerActions as $headerAction) {
+            foreach (Arr::wrap($this->evaluate($headerAction)) as $action) {
+                $this->cachedHeaderActions[$action->getName()] = $this->prepareAction(
+                    $action
+                        ->defaultSize(ActionSize::Small)
```
Comment (zepfietje) [inline packages/infolists/src/Components/Concerns/HasHeaderActions.php]: Probably remove the default size and view as they're not needed.

File: packages/support/resources/views/components/section/index.blade.php (hunk 9 above)
Comment (danharrin) [reply packages/support/resources/views/components/section/index.blade.php:135]: Replaced the class

File: packages/infolists/src/Components/Concerns/HasHeaderActions.php (hunk 10 above)
Comment (danharrin) [reply packages/infolists/src/Components/Concerns/HasHeaderActions.php]: done

File: docs-assets/screenshots/images/light/infolists/layout/section/actions.jpg
Comment (danharrin) [reply docs-assets/screenshots/images/light/infolists/layout/section/actions.jpg:1]: done

File: docs-assets/screenshots/images/light/forms/layout/section/actions.jpg
Comment (danharrin) [reply docs-assets/screenshots/images/light/forms/layout/section/actions.jpg:1]: done

**Pull request merged**